google_cloudsql/mysql/main.tf

```diff
@@ -0,0 +1,130 @@
+/**
+ * # cloudsql-mysql
+ * Creates CloudSQL MySQL Instance.
+ */
+
+locals {
+  default_database_name = "${var.application}-${var.realm}-${var.environment}-${var.instance_version}"
+  database_name         = coalesce(var.custom_database_name, local.default_database_name)
+  tier                  = coalesce(var.tier_override, "db-custom-${var.db_cpu}-${var.db_mem_gb * 1024}")
+
+  default_replica_name = "${local.database_name}-replica"
+  replica_name         = coalesce(var.custom_replica_name, local.default_replica_name)
+
+  ip_addresses = google_sql_database_instance.master.ip_address
+
+  enable_ha = var.realm == "prod" ? true : false
+}
+
+resource "google_sql_database_instance" "master" {
+  name             = local.database_name
+  region           = var.region
+  database_version = var.database_version
+
+  settings {
+    tier = local.tier
+
+    availability_type = var.availability_type
+
+    disk_size = var.disk_size
+    disk_type = "PD_SSD"
+    dynamic "database_flags" {
+      for_each = var.database_flags
+      content {
+        name  = lookup(database_flags.value, "name", null)
+        value = lookup(database_flags.value, "value", null)
+      }
+    }
+
+    backup_configuration {
+      binary_log_enabled = true
+      enabled            = true
+      start_time         = "20:00"
+      location           = "us"
+
+      backup_retention_settings {
+        retained_backups = 30
+        retention_unit   = "COUNT"
+      }
+    }
+
+    ip_configuration {
+      ipv4_enabled    = var.enable_public_ip
+      private_network = var.network
+      require_ssl     = var.ip_configuration_require_ssl
+      dynamic "authorized_networks" {
+        for_each = var.authorized_networks
+        content {
+          expiration_time = lookup(authorized_networks.value, "expiration_time", null)
+          name            = lookup(authorized_networks.value, "name", null)
+          value           = lookup(authorized_networks.value, "value", null)
+        }
+      }
+    }
+
+    maintenance_window {
+      day          = "2"
+      hour         = "16"
+      update_track = var.realm == "prod" ? "stable" : "canary"
+    }
+
+    user_labels = {
+      app         = var.application
+      environment = var.environment
+      realm       = var.realm
+    }
+  }
+
+  lifecycle {
+    prevent_destroy = true
+    ignore_changes  = [settings.0.disk_size]
+  }
+}
+
+resource "google_sql_database_instance" "replica" {
```

Comment:
This is the old way of doing HA - might be good to drop this and change to the new way (which I think is just setting a boolean attribute on the master instance)

Can read in depth at https://cloud.google.com/sql/docs/mysql/high-availability#legacy_mysql_high_availability_option

Reply:
Yeah, I just re-did that stuff a bit. HA is determined when you set `availability_type` to REGIONAL and replicas are a separate thing.